Port Zephyr to Intel S1000

rgundi opened the pull request:
This patchset brings Zephyr support to Intel S1000. The GPIO, UART, I2C, DMA and I2S drivers along with interrupt controllers are currently supported.

rgundi committed:
intel_s1000: Add intel_s1000 SoC

intel_s1000 is an SoC having cavs21_LX6HiFi3_RF3_WB16 as the CPU which
belongs to Xtensa family. This is being used in intel_s1000_crb.

Change-Id: Ic424aa77557bf31024ddbf3f1d76b72a4adb8f66
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
intel_s1000: define memctl_default.S and memerror-vector.S files

Define memctl_default.S and memerror-vector.S files. A reference
could be found in the Xtensa toolchain directories. These are
required for using cavs21_LX6HiFi3_RF3_WB16 Xtensa CPU mainly in
simulator. On boards which have ROM, these would have been already
defined in the ROM. Hence, the contents of these files will be
developed at a later time if required.

Change-Id: Idf52397bb6880c136525e69f47e09defcba7f036
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
intel_s1000: create xt-sim_intel_s1000_defconfig

Create xt-sim_intel_s1000_defconfig to build intel_s1000 SoC
for simulator.

Change-Id: I938bc397ed636b333a14a715a03653de7f3e9351
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
intel_s1000: add intel_s1000_crb board

intel_s1000_crb board makes use of the intel_s1000 SoC. It has
a built-in ROM which gets executed upon applying power. It then
executes the secondary bootloader followed by the FW (like zephyr).

Change-Id: If334c359b4372a56997c3b2e1eb9250e80847f07
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
xtensa: reset-vector.S hack for booting intel_s1000 [REVERTME]

Setting CACHEATTR from _memmap_cacheattr_reset is making
the intel_s1000 SoC get into some unknown state. Removing
it for intel_s1000_crb for now.

Change-Id: Ib44638ef75de6200ef5c2aad55f093a633da864a
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
intel_s1000: uart: configure UART for intel_s1000

intel_s1000 uses DesignWare IP for UART. National Semiconductor
16550 (UART) component specification is followed in this IP.

Change-Id: Ied7df1dc178d55b6dbe71d729d6383ba07274ea4
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
arch: xtensa: set __start as entry point for Xtensa.

__start is deemed the entry point for all architectures in Zephyr.
Accordingly, Xtensa code had to be modified a bit to fall in line
with this convention.

Change-Id: If3ed344721c9f2735378b866662a68d8d5795324
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
intel_s1000: gpio: enable GPIO handling

Enable GPIO handling for intel_s1000. It uses a DesignWare IP.

Change-Id: I522534935e4ef3a56d93aca669f6de961d927481
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
intel_s1000: i2c: enable I2C for intel_s1000

intel_s1000 makes use of DesignWare IP for I2C.

Change-Id: Ie091318c5959b95e1febeb5cefa440f35a6d144b
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
drivers: interrupts: introduce CAVS interrupt logic

CAVS interrupt logic is an intel IP that combines several sources of
interrupt into one line that is then routed to the parent controller.
CAVS stands for "connected Audio, Voice and Speech". This IP supports
4 lines which can have a max of 32 interrupts each.

Change-Id: Ia6be51428bedf1011d148ae1fc5d4c34252c05da
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
drivers: interrupts: introduce Designware interrupt controller

This interrupt controller is a designware IP that combines several
sources of interrupt into one line that is then routed to the parent
controller.

This implementation supports only the regular irqs with no support
for priority filtering and vectored interrupts. Firqs are also not
supported.

Change-Id: I8bdf6f8df4632b6d7e8a3ba9a77116771d034a48
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
xtensa: intel_s1000: implement interrupt mechanism

intel_s1000 has multiple levels of interrupts consisting of core, CAVS
Logic and designware interrupt controller. This patchset modifies
the regular gen_isr mechanism to support these multiple levels.

Change-Id: I0450666d4e601dfbc8cadc9c9d8100afb61a214c
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
intel_s1000: tests: introduce tests to check features enabled

This patchset carries out basic tests on the features supported
on intel_s1000. Currently, Interrupt mechanism, GPIO handling, I2C
communication and UART prints are illustrated.

Change-Id: I7ea03b5085b7fa8d29635c294038536465a70660
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
xtensa: provide XCC compiler support for Xtensa

This patchset provides Xtensa's xcc compiler support for Xtensa
projects in Cmake. This requires the below environment variables
to be defined aptly. The appropriate xcc license information also
need to be supplied.

ZEPHYR_GCC_VARIANT=xcc
TOOLCHAIN_VER=RF-2015.3-linux
XTENSA_CORE=cavs21_LX6HiFi3_RF3_WB16
XTENSA_SYSTEM=/opt/xtensa/XtDevTools/install/tools/
		RF-2015.3-linux/XtensaTools/config/
XTENSA_BUILD_PATHS=/opt/xtensa/XtDevTools/install/builds/

Change-Id: Ib3c10e8095439b0e32276ff37c00eca8420773ec
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
drivers: dma: introduce Intel CAVS DMA

Introduce the Intel CAVS DMA. This is based out of the DesignWare
DMA IP but the register offset and bits have been changed in some
cases. However, the fundamental definition for the register field
has not been changed. Hence the registers begin with "DW_" to
indicate the Designware origin.

This driver currently supports the single block mode and linked list
multi-block mode. Scatter-Gather is not supported.

Change-Id: I33a8ed5141d9236167de50e14d3d407e95d6f553
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
boards: intel_s1000_crb: enable DMA

Enable the CAVS DMA on intel_s1000. Also, introduce a test to
validate the DMA.

Change-Id: I2ff233c45cfd8aea55e254d905350a666aa649a0
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
intel_s1000: scripts: debug, debugserver and flash scripts

This patchset creates Debug, Debugserver and Flash scripts
ensuring support in the ZephyrBinaryRunner mode.

Change-Id: Ib4f7820b1c6a045bd67cf4a031be99cf61e65eca
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
drivers: i2s: introduce CAVS I2S

Implements the driver for Intel CAVS I2S. Only Playback
is currently supported.

Change-Id: I7b816f9736dc35e79a81d3664d6405dc0aac15b4
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
boards: intel_s1000: enable I2S

Enable the CAVS I2S on intel_s1000. Also, introduce a test to
validate I2S in ping-pong and normal modes.

Change-Id: Iea78e22cedc7724c1c2e821c68d7030cc0a65047
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

SavinayDharmappa committed:
driver: usb: enable usb2.0 on intel_s1000

Change-Id: I59091c465e84f1d6ba64433b7320182d83ff9589
Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

SavinayDharmappa committed:
test: board: intel_S1000: Add test application for HID

patch add a hid test application which sends continues
reports to the host

Change-Id: I3a638caded629b502b4cfc6c7bdcba0b67428350
Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

SavinayDharmappa committed:
tests: boards: intel_s1000_crb: select application based on config

Add support to enable test application selectively based
on defconfig

Change-Id: I8111750ff099edd95ffa22c6e312999c801678ea
Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

rgundi committed:
doc: brief description on intel_s1000 usage

A brief description about intel_s1000 and ways to connect,
flash, debug etc.

Change-Id: I459c64d8d9574bc3511da7f983d68e044cacc1d0
Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>
Signed-off-by: Anas Nashif <anas.nashif@intel.com>

SavinayDharmappa committed:
dts: xtensa: Add device tree support for xtensa

Change-Id: I309bc50c6b575caa84fbc7ab98cc9890771b4274
Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa committed:
dts: interrupt_controller: Add dts support for DesignWare controller

Add dts support for multilevel DW interrupt controller

Change-Id: Ia16d6870bd3a46fca933c906aedc6ba78ed5131a
Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa committed:
dts: i2c: Add dts support for i2c

Change-Id: I9104f5230a4efbf1b6979e146e7ea73a76891947
Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

mbolivar commented on boards/xtensa/intel_s1000_crb/board.cmake:
```diff
@@ -1,2 +1 @@
-set(DEBUG_SCRIPT xt-debug.sh)
```
Please combine these into a single commit. The older shell script-based way of working is not supported anymore.

rgundi replied:
Done.

andyross commented on arch/xtensa/soc/intel_s1000/linker.ld:
```diff
@@ -0,0 +1,537 @@
+_memmap_seg_srom2_start = 0xbefe0400;
+_memmap_seg_srom2_max   = 0xbefe0580;
+_memmap_seg_srom4_start = 0xbefe0800;
+_memmap_seg_srom4_max   = 0xbefe0978;
```
Sorry, who is using all these symbols?  This is a staggering amount of typing for what AFAICT is a bunch of untested numbers copied out of a book somewhere.  That sounds dangerous.  Do we really need all this segment boundaries in the linker script if we aren't putting anything there?

andyross commented on arch/xtensa/core/startup/reset-vector.S:
```diff
@@ -433,10 +433,16 @@ _xtos_mpu_attribs:
 		|| XCHAL_HAVE_XLT_CACHEATTR \
 		|| (XCHAL_HAVE_PTP_MMU && XCHAL_HAVE_SPANNING_WAY)
+/* FixMe: Setting CACHEATTR from _memmap_cacheattr_reset is making the
+ * Intel_S1000 SoC get into some unknown state. Removing them only for
```
Dumb question: are you sure you got the value of memmap_cacheattr_reset right in the linker script?  That seems like a weird place to have to store it, and it's reliant on developers typing them correctly instead of emitting it as part of the XCHAL_* stuff.

andyross commented on drivers/i2s/i2s_cavs.h:
```diff
@@ -0,0 +1,137 @@
+extern "C" {
+#endif
+
+struct ssp {
```
Three-letter struct names are likely to collide and don't belong in drivers.  Maybe "i2s_cavs_ssp" instead?

rgundi replied:
sure. will make the change.

andyross commented on drivers/i2s/i2s_cavs.h:
```diff
@@ -0,0 +1,137 @@
+	u32_t ssc3;		/* 0x48 - Command / Status 3 */
+	u32_t ssioc;		/* 0x4C - IO Control */
+};
+
```
More broadly: why is this stuff in a header at all?  It defines internal data that is only used by or included in one other file, and exposes no APIs for use by other parts of the system.  Just put this in the C file.

rgundi replied:
This is more of a personal choice, i feel. I wanted to keep some things in the header file to keep the size of the source file reasonable. Also, i wasn't sure if the things that i put in the header file would not be used anywhere else (except the corresponding .c). That said, i can merge the .c an .h files if you insist.

andyross commented on drivers/i2s/i2s_cavs.c:
```diff
@@ -0,0 +1,791 @@
+	memset(&blk_cfg, 0, sizeof(blk_cfg));
+	blk_cfg.block_size = blk_size;
+	blk_cfg.source_address = (u32_t)src;
+	blk_cfg.dest_address = (u32_t)dst;
```
Style; doing this with a struct literal allows the compiler to generate better code and is more readable:

    blk_cfg = {
       .block_size = blk_size,
       .source_address = (u32_t)src,
       .dest_address = (u32_t)dst,
    }

rgundi replied:
sure. will make the change.

andyross commented on drivers/i2s/i2s_cavs.c:
```diff
@@ -0,0 +1,791 @@
+	/* enable port */
+	ssp->ssc0 |= SSCR0_SSE;
+
+	/* enable DMA */
```
So... start_dma() doesn't enable DMA?  Is that racy?   Seems like this needs a comment explaining why this is OK.  Maybe "start_dma()" should be renamed "setup_dma()"?

rgundi replied:
Well.. here the DMA is controlled by the peripheral. So, even if the DMA starts, unless the peripheral enables the DMA, the DMA will not complete. I can put in a comment there.

andyross commented on drivers/i2s/i2s_cavs.c:
```diff
@@ -0,0 +1,791 @@
+#define CAVS_SSP_WORD_PER_FRAME_MAX   16
+
+struct queue_item {
+	void *mem_block;
```
Any reason you can't use a k_queue for this?  Among other things, I don't see any synchronization of the I2C API here, and a k_queue would give you that for free.

rgundi replied:
Guess a k_queue should work here. I don't need any synchronization though. Will give the struct a try and see if things work fine.

andyross reviewed (commented):
Some random whining but nothing that I think should block merge by itself.

I do note the lack of asm2 integration.  Was that not working out with xcc?  Any particular problems?

lpereira commented on drivers/dma/dma_cavs.c:
```diff
@@ -0,0 +1,454 @@
+			 struct dma_config *cfg)
+{
+	struct dw_dma_dev_data *const dev_data = DEV_DATA(dev);
+	struct dma_chan_data *chan_data = &dev_data->dw_pdata.chan[channel];
```
Please check if channel is within bounds before taking a pointer to the chan array.

rgundi replied:
ok.

lpereira commented on drivers/dma/dma_cavs.c:
```diff
@@ -0,0 +1,454 @@
+	return *((volatile u32_t*)(dma_base + reg));
+}
+
+static void dw_dma_isr(void *arg)
```
Is this port in particular capable of SMP? If so, we might want to copy the DMA callbacks to local variables before looping through blocks/tfrs.

rgundi replied:
Currently this port doesn't have SMP but we may need to support it in future. DMA callbacks are already copied to local variables. Can you let me know what is not copied here?

lpereira commented on drivers/i2s/i2s_cavs.c:
```diff
@@ -0,0 +1,791 @@
+	struct i2s_config cfg;
+	struct ring_buf mem_block_queue;
+	void *mem_block;
+	bool last_block;
```
With the embedded structs here, will having a boolean in the middle create a unnecessary hole due to alignment?

rgundi replied:
Well, not sure if a hole is created here. What is the size of bool in zephyr?

lpereira commented:
I've checked most of these patches already, and there are just a few minor points that I'd like to have fixed before merging this.

BTW, thanks for using DT in an Xtensa port.

rgundi commented on arch/xtensa/soc/intel_s1000/soc.c:
```diff
@@ -0,0 +1,234 @@
+	 */
+	irq_enable_next_level(dev_cavs, CAVS_IRQ_NUMBER(irq));
+
+	switch (INTR_CNTL_IRQ_NUM(irq)) {
```
I initially kept it thinking i may need it for something in future. Haven't used it so far and i don't think i need it anymore. Will remove it. Will reintroduce it in future, if needed.

rgundi commented on arch/xtensa/soc/intel_s1000/soc.c:
```diff
@@ -0,0 +1,234 @@
+	 * is in DW interrupt controller
+	 */
+
+	switch (INTR_CNTL_IRQ_NUM(irq)) {
```
will remove this too.

rgundi commented on drivers/dma/dma_cavs.c:
```diff
@@ -0,0 +1,454 @@
+{
+	const struct dw_dma_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	struct dw_dma_dev_data *const dev_data = DEV_DATA(dev);
+	struct dma_chan_data *chan_data = &dev_data->dw_pdata.chan[channel];
```
ok

rgundi commented on tests/boards/intel_s1000_crb/src/test_hid.c:
```diff
@@ -0,0 +1,122 @@
+}
+
+K_THREAD_DEFINE(hid_thread_id, STACKSIZE, hid_thread, NULL, NULL, NULL,
+		PRIORITY, 0, K_NO_WAIT);
```
This is the way it has been done for all the tests for peripherals in S1000. Just keeps "main" clean and it's easier to add and remove tests (via CMakeLists.txt).

rgundi commented on tests/boards/intel_s1000_crb/prj.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_USB=y
+CONFIG_USB_DEVICE_STACK=y
+CONFIG_USB_DEVICE_HID=y
```
YAML descriptor and ztest integration will happen in a future patchset.

rgundi commented on dts/xtensa/intel_s1000.dtsi:
```diff
@@ -0,0 +1,44 @@
+		reg = <0xbe000000 0x300000>;
+	};
+
+	soc {
```
This is not fully complete. We haven't finished migrating all drivers to DTS. This will also happen separately driver by driver in future patchsets.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -189,13 +201,18 @@ if(CONFIG_READABLE_ASM)
+if(NOT ${ZEPHYR_TOOLCHAIN_VARIANT} STREQUAL "xcc")
+  zephyr_cc_option(-fno-asynchronous-unwind-tables)
+  zephyr_cc_option(-fno-pie)
+endif()
```
It should be possible to omit the if statement.

rgundi replied:
ok

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -189,13 +201,18 @@ if(CONFIG_READABLE_ASM)
-zephyr_compile_options_ifdef(CONFIG_STACK_CANARIES -fstack-protector-all)
+if(NOT ${ZEPHYR_TOOLCHAIN_VARIANT} STREQUAL "xcc")
+  zephyr_compile_options_ifdef(CONFIG_STACK_CANARIES -fstack-protector-all)
+endif()
```
What the CMake code was originally expressing here was correct.

If the user has enabled STACK_CANARIES, but the toolchain does not support -fstack-protector-all, then the build system should not silently drop the protection.

SebastianBoe commented on CMakeLists.txt:
```diff
@@ -220,12 +237,12 @@ if(CMAKE_C_COMPILER_ID STREQUAL "Clang")
+      -Wno-unused-but-set-variable
+      -fno-reorder-functions
+      -fno-defer-pop
+    )
```
It should be possible to omit the if statement.

rgundi replied:
ok

SebastianBoe commented on cmake/compiler/xcc.cmake:
```diff
@@ -0,0 +1,93 @@
+  list(APPEND isystem_include_flags -isystem ${isystem_include_dir})
+endforeach()
+set(CMAKE_REQUIRED_FLAGS -nostartfiles -nostdlib ${isystem_include_flags} -Wl,--unresolved-symbols=ignore-in-object-files)
+string(REPLACE ";" " " CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS}")
```
I would recommend organizing this as a "gcc-like" toolchain, so having COMPILER set to gcc, and modifying gcc.cmake to be generic enough to support xcc.

SebastianBoe replied:
No, the code must be maintainable before it goes into master.

Or else the technical debt is pushed to the next contributor to gcc.cmake, instead of where it rightly belongs.

I am aware there are differences, gcc.cmake must be made generic enough to not require the changes made.

rgundi replied:
ok. i have come up with something. Please review once i push the changes.

rgundi commented on arch/xtensa/core/startup/reset-vector.S:
```diff
@@ -24,8 +24,8 @@
-	.global	_ResetVector
-_ResetVector:
+	.global	__start
+__start:
```
Well.. on S1000, the "make debug" and "make flash" commands download the ELF file on to the beginning of SRAM (address is 0xBE000000). Upon Power on, ROM code gets executed and jumps to SRAM location 0xBE000000. So, the first piece of code that we want to be executed in SRAM needs to be placed at 0xBE000000. Since there is already a different __start in the Xtensa code, that was getting placed at that location thereby causing issues. Hence _ResetVector had to be renamed to __start to ensure it gets included in the right area.

rgundi commented on arch/xtensa/core/startup/reset-vector.S:
```diff
@@ -433,10 +433,16 @@ _xtos_mpu_attribs:
+ * Intel_S1000 SoC get into some unknown state. Removing them only for
+ * Intel_S1000_CRB
+ */
+#ifndef CONFIG_BOARD_INTEL_S1000_CRB
```
I understand your point but i am unsure how to implement it. Can we take this up later?

dbkinder commented on boards/xtensa/intel_s1000_crb/doc/intel_s1000.rst:
```diff
@@ -0,0 +1,216 @@
+Setting up UART
+===============
+
+We use a "FT232RL FTDI USB To TTL Serial Converter Adapter Module" to
```
Saying "We use" is an unexpected shift.  Maybe instead, say, "We recommend using" here.

dbkinder commented on boards/xtensa/intel_s1000_crb/doc/intel_s1000.rst:
```diff
@@ -0,0 +1,216 @@
+
+We use a "FT232RL FTDI USB To TTL Serial Converter Adapter Module" to
+tap the UART data. The J8 Header on S1000 CRB is dedicated for UART.
+Please do the connection between the J8 header and UART chip as shown below.
```
Delete the "please":

    Connect the J8 header and UART chip as shown below:

dbkinder commented on boards/xtensa/intel_s1000_crb/doc/intel_s1000.rst:
```diff
@@ -0,0 +1,216 @@
+| GND        | 10        |
++------------+-----------+
+
+We use minicom for monitoring the UART data. Follow the below steps
```
Instead:

    Use ``minicom`` or another terminal emulator to monitor the UART data
    by following these steps:

dbkinder commented on boards/xtensa/intel_s1000_crb/doc/intel_s1000.rst:
```diff
@@ -0,0 +1,216 @@
+Using JTAG
+==========
+
+We use a flyswatter2 debugger to connect to S1000 CRB. The pinouts for
```
Instead:

    For debugging, you can use a flyswatter2 to connect to the S1000 CRB.

dbkinder commented on boards/xtensa/intel_s1000_crb/doc/intel_s1000.rst:
```diff
@@ -0,0 +1,216 @@
+
+References
+**********
+
```
If you intended for these links to show up in the generated output, add this directive before them:

    .. target-notes::

and a blank line.

SavinayDharmappa commented:
@galak could you review dts support for xtensa ?

mbolivar commented on scripts/support/runner/intel_s1000.py:
```diff
@@ -0,0 +1,138 @@
+        # Note that XTOCD always fails the first time. It has to be
+        # relaunched the second time to work.
+        self.popen_ignore_int(server_cmd)
+        time.sleep(3)
```
So this board needs to wait 6 seconds total for the command to fail and then be re-run before a flash can succeed? Sorry to hear it :(, sounds like silicon issues.

That's funny and unfortunate, and no problem for other runners, but I am not sure if popen_ignore_int is the right API to be used here. Do you really need to ignore SIGINT? Or is there a missing superclass call for creating a popen object that doesn't require a new process group?

rgundi replied:
No idea if it is a silicon issue but this is how the board is being used. Please note there will be some verbose prints coming on the console, so 6 seconds are tolerable.

I am using popen_ignore_int so that i can move over to the next command. If i use check_call, it will be blocking. As you can see, multiple commands have to be entered and automation is not possible if they are blocking calls.

mbolivar commented on scripts/support/runner/intel_s1000.py:
```diff
@@ -0,0 +1,138 @@
+
+        # The below statement will consume the "^C" keypress ensuring
+        # the python main application doesn't exit. This is important
+        # since ^C in gdb means a "halt" operation.
```
that's what the popen_ignore_int call is meant to handle, except it is cross-platform, unlike this call which only works on POSIX systems... why use that above but not here?

rgundi replied:
If i use popen_ignore_int call here, the behaviour becomes extremely weird. if i press Enter, it alternates between gdb and the shell. That's why i had to come up with this thing.

mbolivar commented on scripts/support/runner/intel_s1000.py:
```diff
@@ -0,0 +1,138 @@
+
+        # Note that XTOCD always fails the first time. It has to be
+        # relaunched the second time to work.
+        self.popen_ignore_int(server_cmd)
```
similar questions to above

rgundi commented on CMakeLists.txt:
```diff
@@ -99,6 +104,10 @@ zephyr_compile_options(
 )
 
+if(${ZEPHYR_TOOLCHAIN_VARIANT} STREQUAL "xcc")
+  zephyr_compile_options(-fms-extensions)
```
The XCC toolchain is derived from an older version of gcc and it does not understand "anonymous structs/unions". Guess it still follows the C99 standard. This flag enables that capability.

rgundi commented on CMakeLists.txt:
```diff
@@ -119,10 +128,13 @@ zephyr_ld_options(
 endif()
 
+if(NOT ${ZEPHYR_TOOLCHAIN_VARIANT} STREQUAL "xcc")
+  zephyr_ld_options(-no-pie)
```
xcc does not support this flag.

I know that zephyr_ld_options is supposed to test and omit a flag if it is not supported but i saw a few failures initially and was not really sure of the impact. I had then decided to omit it for XCC. Will revert this change now.

rgundi commented on CMakeLists.txt:
```diff
@@ -99,6 +104,10 @@ zephyr_compile_options(
 )
 
+if(${ZEPHYR_TOOLCHAIN_VARIANT} STREQUAL "xcc")
+  zephyr_compile_options(-fms-extensions)
```
should i employ a method similar to the one adopted for OPTIMIZE_FOR_SIZE_FLAG flag? i.e. define a new flag, something like FMS_EXTENSIONS in toolchain build script and then include it under zephyr_compile_options in the toplevel CMakeLists.txt.
Another option is to use zephyr_cc_option directly. It will fail for gcc but work for xcc. Let me know your thoughts.

SebastianBoe replied:
Hi, I don't think -fms-extensions will fail for gcc actually.

I believe the cleanest solution will be to do

  list(APPEND TOOLCHAIN_C_FLAGS -fms-extensions)

from the toolchain build script.

rgundi replied:
Well.. that was my preferred way as well in the beginning when i started providing this support. It just doesn't work. It just doesn't seem to take effect at all. That's the reason why i modified the toplevel CMakeLists.txt.

SebastianBoe replied:
Right, so the reason TOOLCHAIN_C_FLAGS is taking effect on 'arm', but not on 'xtensa' is that arm is doing

zephyr_compile_options(
  ${TOOLCHAIN_C_FLAGS}
  )

I suspect that this is undesirable, the toplevel CMakeLists.txt should be doing

zephyr_compile_options(
  ${TOOLCHAIN_C_FLAGS}
  )

so that it is not architecture-dependent whether toolchain flags are used or not.

Anyway, for this particular PR we need to get rid of

+if(${ZEPHYR_TOOLCHAIN_VARIANT} STREQUAL "xcc")

in the toplevel CMakeLists.txt file.

This can be done by copying what 'arm' is doing, and executing

 zephyr_compile_options(${TOOLCHAIN_C_FLAGS})

in arch/xtensa/CMakeLists.txt

or by adding TOOLCHAIN_C_FLAGS in the toplevel CMakeLists.txt and removing it from arch/{arm,posix}.

rgundi replied:
Thanks for the info. I have modified the toplevel CMakeLists.txt and it is working fine.

rgundi commented on CMakeLists.txt:
```diff
@@ -86,6 +86,11 @@ else()
+if(${ZEPHYR_TOOLCHAIN_VARIANT} STREQUAL "xcc")
+  string(REPLACE "-Og" "-O0" OPTIMIZE_FOR_SIZE_FLAG "${OPTIMIZE_FOR_SIZE_FLAG}")
+endif()
+
```
Agreed.

rgundi commented:
@andyross @lpereira I have addressed most of your queries in the latest patchset. Please have a look. @andyross Xtensa_Asm2 is not yet enabled for S1000 but this is something i believe should be tried out later.
@SebastianBoe I have attempted implementing your asks. Please have a look. Haven't moved the -fms-extensions thing though. Needed your input there.
@dbkinder I have modified the doc as per your suggestion. Only the "contact Intel" thing is still pending since things are still being worked out. I will update that part once things are more concrete.
@mbolivar No new changes in this patchset. I have answered your questions inline.

SebastianBoe commented on cmake/compiler/gcc.cmake:
```diff
@@ -1,10 +1,18 @@
+elseif("${ZEPHYR_TOOLCHAIN_VARIANT}" STREQUAL "xcc")
+  set(cc "xcc")
+  set(c++ "xc++")
+endif()
```
gcc.cmake should not make references to individual toolchains unless necessary.

In this particular case it is not necessary because one could have an overridable default for 'cc' and 'c++'.

set_ifndef(cc gcc)

(And then set(cc xcc) in xcc.cmake)

rgundi replied:
Will make this change.

SebastianBoe commented on cmake/compiler/gcc.cmake:
```diff
@@ -1,10 +1,18 @@
+if("${ZEPHYR_TOOLCHAIN_VARIANT}" STREQUAL "zephyr")
+  set(cc "gcc")
```
In the zephyr build system it would be more common to have an uppercase name for 'cc'.

rgundi replied:
Will make this change.

SebastianBoe commented on cmake/compiler/gcc.cmake:
```diff
@@ -50,56 +58,76 @@ foreach(file_name include include-fixed)
 endforeach()
 
-include(${ZEPHYR_BASE}/cmake/gcc-m-cpu.cmake)
+if("${ZEPHYR_TOOLCHAIN_VARIANT}" STREQUAL "zephyr")
```
This is incorrect.

We want to append -mthumb for more toolchain's than 'zephyr', for instance 'gccarmemb' would also use this build script code.

This if statement must be removed.

rgundi replied:
Will rearrange the "ifs" ensuring the else part falls here.

SebastianBoe commented on cmake/compiler/gcc.cmake:
```diff
@@ -50,56 +58,76 @@ foreach(file_name include include-fixed)
+elseif("${ZEPHYR_TOOLCHAIN_VARIANT}" STREQUAL "xcc")
+  # Include XCC standard libraries so that users used to Xplorer IDE can port
+  # their code easily
+  set(TOOLCHAIN_INCLUDES
```
Is it necessary for xcc to diverge wrt. discovering libc?

Specifically, does xcc support '--print-libgcc-file-name' ?

If so, it should be used to discover the LIB_INCLUDE_DIR where the 'gcc' toolchain lib is located.

rgundi replied:
xcc supports --print-libgcc-file-name but it does not need to look up libgcc.a. I can see that i am able to compile and link without that. So, have not included that for xcc.

SebastianBoe commented on cmake/compiler/gcc.cmake:
```diff
@@ -50,56 +58,76 @@ foreach(file_name include include-fixed)
-LIST(APPEND LIB_INCLUDE_DIR "-L\"${LIBGCC_DIR}\"")
-LIST(APPEND TOOLCHAIN_LIBS gcc)
+  LIST(APPEND TOOLCHAIN_LIBS gcc)
+  LIST(APPEND TOOLCHAIN_LIBS hal)
```
HAL libraries, their header file include paths, and their lib include directories should not be added by the toolchain.

If they are added by the toolchain it will not be possible to opt-out of using the hal.

Instead, they should be added through Kconfig.

This is an example of the same situation that xcc's 'hal' is in:

https://github.com/zephyrproject-rtos/zephyr/blob/master/ext/hal/qmsi/CMakeLists.txt#L12

It is a toolchain vendor-supplied binary 'hal' library with an include directory for header files and an include directory for the binary lib itself. This library must be organized in the same manner if possible.

rgundi replied:
I had a relook at what i am doing for XCC in this file and i see that only below 2 lines are required and i can omit everything else.
  LIST(APPEND TOOLCHAIN_LIBS gcc)
  LIST(APPEND TOOLCHAIN_LIBS hal)

The other things included seem to be for facilitating the users used to Xplorer IDE (XCC standard libraries) to port their code easily.

rgundi replied:
Had a look at the Kconfig thing. I think it may not be applicable in this case as there is nothing configurable here. If you plan to use XCC for your build, you just get a set of preconfigured things. If you want to use GCC, then apparently this HAL is not required.

SebastianBoe commented:
"@SebastianBoe I have attempted implementing your asks. Please have a look. Haven't moved the -fms-extensions thing though. Needed your input there."

@rgundi : I have reviewed the new revision. Great that you were able to get rid of the duplicated gcc.cmake.

rgundi commented:
@SebastianBoe : I have addressed all your comments except for the Kconfig thing. Please review them while i have a look at the Kconfig thing.

rgundi commented:
@andyross @lpereira @dbkinder @mbolivar  : Please review the patchset and check if your comments have been addressed.

dbkinder commented on boards/xtensa/intel_s1000_crb/doc/intel_s1000.rst:
```diff
@@ -0,0 +1,216 @@
+=============
+
+The Xtensa 'toolchain'_ i.e. XCC is required to build this port. This needs a
+license and is available for Linux and Windows. Please contact Intel for getting
```
Again, there's no information here about how to contact Intel or how to get a license.  When I visit the login page mentioned below (https://xpg.cadence.com/login/gen/ten4genlogin.html) there's no mention there either of how to get a login/license.  We need a link to information about how to do this.

dbkinder commented on boards/xtensa/intel_s1000_crb/doc/intel_s1000.rst:
```diff
@@ -0,0 +1,248 @@
+
+The Xtensa processor in the S1000 board (cavs21_LX6HiFi3_RF3_WB16) is not
+supported by the stock RF-2015.3 (Xplorer-6.0.3) tools installed above. Download
+the Cadence XCC build for cavs21_LX6HiFi3_RF3_WB16 (a .tgz file) by contacting
```
again, how do you "contact Intel"?  How do you download the tgz file?  (no link is provided).

dbkinder reviewed (changes requested):
Thanks for the updates.  We still need some information about how to get a login and license for downloading the Cadence files.  Saying "Contact Intel" isn't enough.

rgundi commented:
@dbkinder : Currently, there are folks at Intel who are talking to Cadence to get access to the toolchain, login credentials, licenses etc. It's taking time since it's business/legal discussions. I have mentioned this in the doc now. Will update the doc once we have this data. Hope this info is ok for now.

nashif commented:
@rgundi, @dbkinder

> Thanks for the updates. We still need some information about how to get a login and license for downloading the Cadence files. Saying "Contact Intel" isn't enough.

should be enough for now, those have the board will be in contact with the supplier who can give them access to license and toolchain needed.

dbkinder commented on boards/xtensa/intel_s1000_crb/doc/intel_s1000.rst:
```diff
@@ -0,0 +1,253 @@
+
+The Xtensa 'toolchain'_ i.e. XCC is required to build this port. This needs a
+license and is available for Linux and Windows. Intel is currently working with Cadence
+to make the license available for wider audience. Watch this space for more updates.
```
Given Anas's comment, maybe better than saying "Watch this space..." would be:

    For now, please contact your board supplier for licensing information and tool chain access.

rgundi replied:
done.

dbkinder commented on boards/xtensa/intel_s1000_crb/doc/intel_s1000.rst:
```diff
@@ -0,0 +1,253 @@
+========================
+
+To install the Xplorer software, log on to https://xpg.cadence.com/login/gen/ten4genlogin.html.
+using your browser. Intel is currently working with Cadence to make a login name and password
```
Since you just said this, I'd remove the two sentences, "Intel is currently..." and "Stay tuned...".

rgundi replied:
done.

dbkinder reviewed (approved):
Thanks!

lpereira commented on include/toolchain/xcc.h:
```diff
@@ -38,4 +38,140 @@
+						unsigned long int)) { \
+		__ret__ = __builtin_uaddl_overflow(a, b, output); \
+	} else { \
+		/* assume overflow for all other types */ \
```
Might be better to just use a BUILD_ASSERT_MSG(0, "__builtin_add_overflow not defined for these types") here: failing at build time is always better than runtime.

lpereira commented on include/toolchain/xcc.h:
```diff
@@ -38,4 +38,140 @@
+		__ret__ = __builtin_uaddl_overflow(a, b, output); \
+	} else { \
+		/* assume overflow for all other types */ \
+		__ret__ = 1; \
```
Might be a better idea to have an assertion and a CODE_UNREACHABLE here.  I don't know if a BUILD_ASSERT_MSG() would work here, but if it does, it's always better to fail at compile time rather than at runtime.

lpereira commented on include/toolchain/xcc.h:
```diff
@@ -38,4 +38,140 @@
+	__ret__;\
+})
+
+#define __builtin_sadd_overflow(a, b, output) ({ \
```
Oh, xcc doesn't even have these?  You can maybe simplify this by using a macro such as this:

```C
#define _xcc_bin_op_overflow(a, b, small_type, overflow_type, operator, output) \
  ({ \
          overflow_type result = a operator b; \
          *output = (small_type)result; \
          *output != (overflow_type)(small_type)result; \
   })
```

And then defining, for instance, `__builtin_sadd_overflow` to `_xcc_bin_op_overflow(a, b, i32_t, i64_t, +, output)`, and so on.  Better to use the explicit types as well (`i32_t` vs `int`), because sometimes, `int` and `long int` are the same, and you want the wider type to capture the overflow.  The same macro can be used for multiplication as well.

I don't know, however, how good XCC is here to know that you're only interested in knowing that the operation overflowed, so you might want to look at alternative methods of detecting overflow if the generated code is terrible when promoting 32 to 64 bit.

lpereira reviewed (commented):
The patches to add __builtin_add_overflow() and __builtin_mul_overflow() could use some refactoring.  They might also not work (I'm not familiar with the ABI used in S1000, so I don't know if `long int` and `int` are the same width), so here are some recommendations.

dbkinder reviewed (approved):
LGTM, thanks!

rgundi commented:
@lpereira : As discussed, i have bypassed the overflow check for xcc. Will revisit it separately.

rgundi commented:
@lpereira , @mbolivar , can you please review and approve this PR?

mbolivar commented on cmake/flash/CMakeLists.txt:
```diff
@@ -1,5 +1,8 @@
-assert_not(FLASH_SCRIPT "FLASH_SCRIPT has been removed; use BOARD_FLASH_RUNNER")
-assert_not(DEBUG_SCRIPT "DEBUG_SCRIPT has been removed; use BOARD_DEBUG_RUNNER")
+# FIXME: This will be fixed once the scripts are written for xcc
```
It seems unnecessary to put this here since the scripts are already written.  They're just added later in the series.

Why not rebase fbc49163a ("intel_s1000: scripts: debug, debugserver and flash scripts") to just before this commit, so that Zephyr's history never permits setting the deprecated variables?

rgundi replied:
Done. Thanks for the suggestion.

lpereira commented on include/toolchain/xcc.h:
```diff
@@ -36,4 +36,12 @@
+/* TODO: XCC doesn't define the below macros which are useful for checking
+ * overflows. This needs to be fixed.
+ */
+#define __builtin_add_overflow(a, b, output) 0
```
The operation must be performed still, otherwise `output` won't be set and things won't work as expected:

```C
#define __builtin_add_overflow(a, b, output)   ({ *output = (a) + (b); 0; })
#define __builtin_mull_overflow(a, b, output)   ({ *output = (a) * (b); 0; })
```

rgundi replied:
done.

rgundi commented:
@lpereira , @mbolivar : I have incorporated your suggestions. Please review.

nashif commented:
recheck
